perfkitbenchmarker/flag_util.py

```diff
@@ -0,0 +1,198 @@
+# Copyright 2015 Google Inc. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Utility functions for working with user-supplied flags."""
+
+import logging
+import re
+
+from perfkitbenchmarker import flags
+
+INTEGER_GROUP_REGEXP = re.compile(r'(\d+)(-(\d+))?$')
+
+
+class IntegerList(object):
+  """An immutable list of nonnegative integers.
+
+  The list contains either single integers (ex: 5) or ranges (ex:
+  8-12). The list can include as many elements as will fit in
+  memory. Furthermore, the memory required to hold a range will not
+  grow with the size of the range.
+
+  Make a list with
+    lst = IntegerList(groups)
+
+  where groups is a list whose elements are either single integers or
+  2-tuples holding the low and high bounds of a range
+  (inclusive). (Ex: [5, (8,12)] represents the integer list
+  5,8,9,10,11,12.)
+
+  """
+
+  def __init__(self, groups):
+    self.groups = groups
+
+    length = 0
+    for elt in groups:
+      if isinstance(elt, int) or isinstance(elt, long):
+        length += 1
+      if isinstance(elt, tuple):
+        length += elt[1] - elt[0] + 1
+
+    self.length = length
+
+  def __len__(self):
+    return self.length
+
+  def __getitem__(self, idx):
+    if not isinstance(idx, int):
+      raise TypeError()
+    if idx < 0 or idx >= self.length:
+      raise IndexError()
+
+    group_idx = 0
+    while idx > 0:
+      group = self.groups[group_idx]
+
+      if not isinstance(group, tuple):
+        group_idx += 1
+        idx -= 1
+      else:
+        group_len = group[1] - group[0] + 1
+        if idx >= group_len:
+          group_idx += 1
+          idx -= group_len
+        else:
+          return group[0] + idx
+
+    if isinstance(self.groups[group_idx], tuple):
+      return self.groups[group_idx][0]
+    else:
+      return self.groups[group_idx]
+
+  def __iter__(self):
+    group_idx = 0
+    in_group_idx = 0
+
+    while group_idx < len(self.groups):
```

Comment:
All of the index tracking muddles the intent a bit here. Could you simplify by iterating directly? Something like (untested):

```{python}
for group in self.groups:
  if (single value):
    yield group
  else:
    low, high = group
    for x in xrange(low, high+1):
      yield x
```